sql/lease.go

```diff
@@ -0,0 +1,511 @@
+	// has less range than time.Time.
+	expiration int64
+	refcount   int
+}
+
+func (s *LeaseState) String() string {
+	return fmt.Sprintf("%d:%d", s.Version, s.expiration)
+}
+
+// Expiration returns the expiration time of the lease.
+func (s *LeaseState) Expiration() time.Time {
+	return time.Unix(0, s.expiration)
+}
+
+// Refcount returns the reference count of the lease.
+func (s *LeaseState) Refcount() int {
+	return s.refcount
+}
+
+// LeaseStore implements the operations for acquiring and releasing leases and
+// publishing a new version of a descriptor. Exported only for testing.
+type LeaseStore struct {
+	db     client.DB
+	clock  *hlc.Clock
+	nodeID uint32
+}
+
+// jitteredLeaseDuration returns a randomly jittered duration from the interval
+// [0.75 * leaseDuration, 1.25 * leaseDuration].
+func (s LeaseStore) jitteredLeaseDuration() int64 {
+	return int64(float64(leaseDuration) * (0.75 + 0.5*rand.Float64()))
+}
+
+// Acquire a lease on the most recent version of a table descriptor.
+func (s LeaseStore) Acquire(tableID ID, minVersion uint32) (*LeaseState, error) {
+	lease := &LeaseState{}
+	lease.expiration = s.clock.Now().WallTime + s.jitteredLeaseDuration()
+
+	err := s.db.Txn(func(txn *client.Txn) error {
+		p := planner{txn: txn, user: security.RootUser}
+
+		const getDescriptor = `SELECT descriptor FROM system.descriptor WHERE id = %d`
+		sql := fmt.Sprintf(getDescriptor, tableID)
+		values, err := p.queryRow(sql)
+		if err != nil {
+			return err
+		}
+		if values == nil {
+			return fmt.Errorf("table ID %d not found", tableID)
+		}
+		if err := proto.Unmarshal([]byte(values[0].(parser.DBytes)), &lease.TableDescriptor); err != nil {
+			return err
+		}
+
+		if err := lease.Validate(); err != nil {
+			return err
+		}
+		if lease.Version < minVersion {
+			return fmt.Errorf("version %d of table %d does not exist yet", minVersion, tableID)
+		}
+
+		const insertLease = `INSERT INTO system.lease (descID, version, nodeID, expiration) ` +
+			`VALUES (%d, %d, %d, '%s'::timestamp)`
+		sql = fmt.Sprintf(insertLease, lease.ID, lease.Version, s.nodeID,
+			nanosToDTimestamp(lease.expiration))
+		count, err := p.exec(sql)
+		if err != nil {
+			return err
+		}
+		if count != 1 {
+			return fmt.Errorf("%s: unexpected result count: %d", sql, count)
+		}
+		return nil
+	})
+	return lease, err
+}
+
+// Release a previously acquired table descriptor lease.
+func (s LeaseStore) Release(lease *LeaseState) error {
+	return s.db.Txn(func(txn *client.Txn) error {
+		p := planner{txn: txn, user: security.RootUser}
+
+		const deleteLease = `DELETE FROM system.lease ` +
+			`WHERE (descID, version, nodeID, expiration) = (%d, %d, %d, '%s'::timestamp)`
+		sql := fmt.Sprintf(deleteLease, lease.ID, lease.Version, s.nodeID,
+			nanosToDTimestamp(lease.expiration))
+		count, err := p.exec(sql)
+		if err != nil {
+			return err
+		}
+		if count != 1 {
+			return fmt.Errorf("%s: unexpected result count: %d", sql, count)
```

Comment:
this should be util.Errorf, but also it would be helpful to indicate the expected number.